packages/lit-ssr/src/lib/render-lit-html.ts

```diff
@@ -0,0 +1,550 @@
+      }
+    } else {
+      yield escapeTextContent(String(value));
+    }
+  }
+  yield `<!--/lit-part-->`;
+}
+
+export function* renderTemplateResult(
+  result: TemplateResult,
+  renderInfo: RenderInfo
+): IterableIterator<string> {
+  // In order to render a TemplateResult we have to handle and stream out
+  // different parts of the result separately:
+  //   - Literal sections of the template
+  //   - Defined custom element within the literal sections
+  //   - Values in the result
+  //
+  // This means we can't just iterate through the template literals and values,
+  // we must parse and traverse the template's HTML. But we don't want to pay
+  // the cost of serializing the HTML node-by-node when we already have the
+  // template in string form. So we parse with location info turned on and use
+  // that to index into the HTML string generated by TemplateResult.getHTML().
+  // During the tree walk we will handle expression marker nodes and custom
+  // elements. For each we will record the offset of the node, and output the
+  // previous span of HTML.
+
+  const {ops, attrNames} = getTemplate(result);
+
+  /* The next value in result.values to render */
+  let partIndex = 0;
+  let attrIndex = 0;
+
+  for (const op of ops) {
+    switch (op.type) {
+      case 'text':
+        yield op.value;
+        break;
+      case 'node-part': {
+        const value = result.values[partIndex++];
+        yield* renderValue(value, renderInfo);
+        break;
+      }
+      case 'attribute-part': {
+        const name = attrNames[attrIndex++];
+        const statics = op.strings;
+        let attributeName = op.name;
+        const prefix = attributeName[0];
+        const attributePart = new AttributePart(
+          // Passing null for the element is fine since the directive only gets
+          // PartInfo without the node available in the constructor
+          // eslint-disable-next-line @typescript-eslint/no-explicit-any
+          (null as any) as HTMLElement,
+          name,
+          statics
+        );
+        let value =
+          attributePart.strings === undefined
+            ? result.values[partIndex]
+            : result.values;
+        value = attributePart._resolveValue(value, partIndex);
+        const instance = op.useCustomElementInstance
+          ? getLast(renderInfo.customElementInstanceStack)
+          : undefined;
+        if (prefix === '.') {
+          const propertyName = name.substring(1);
+          // Property should be reflected to attribute
+          const reflectedName = reflectedAttributeName(
+            op.tagName,
+            propertyName
+          );
+          // Property should be set to custom element instance
+          const instance = op.useCustomElementInstance
+            ? getLast(renderInfo.customElementInstanceStack)
+            : undefined;
+          if (instance || reflectedName !== undefined) {
+            if (value !== noChange) {
```

Comment:
don't we need to handle `nothing` here now too?

I think we might want to warn on `noChange` also. Seems like it's probably a mistake.

Reply:
Yeah, I fixed up the `nothing` code paths.

Re: `noChange`, it doesn't seem like an initial value of `noChange` is necessarily a mistake. An async directive that just doesn't want to do anything until later might do that right?